### ISSUE TITLE
False negative with `type` annotation in dataclass in 1.1.338

### ISSUE THREAD

**Issue by rsokl** [microsoft/pyright#6596]
```python
def f(x: type): ...

# pyright 1.1.337: "float" is incompatible with "type"
# pyright 1.1.338: (No error)
f(1.0)  
```

[hydra-zen CI](https://github.com/mit-ll-responsible-ai/hydra-zen/actions/runs/7038046324/job/19154139161?pr=605#step:5:33)

**Edit: the above example was wrong** 

Here is the actual false negative example:

```python
from dataclasses import dataclass

@dataclass()
class A:
    x: type

# pyright 1.1.337: "float" is incompatible with "type"
# pyright 1.1.338: (No error)
A(1)
```

**Comment by erictraut** [comment]
I'm not seeing a false negative with this code. Pyright 1.1.338 emits an error as expected.

Code sample in [pyright playground](https://pyright-play.net/?code=CYUwZgBGAUAeBcEAuBPADiAlPAUBfEAdMTjjAIyEAMmOQA)

Can you please confirm that you're seeing something different?

**Comment by rsokl** [comment]
Will do

**Comment by rsokl** [comment]
It turns out that this the issue is specific to dataclass it seems. Apologies for getting the MRP wrong:

```python
from dataclasses import dataclass

@dataclass()
class A:
    x: type

A(1)
```

[playground link](https://pyright-play.net/?pyrightVersion=1.1.338&code=GYJw9gtgBAJghgFzgYwDZwM4YKYagSwgAcwQFZEV0sAoGgAXiTUwwAoBKGlrKAQQBcNKCKgAPAVAQBPItjp82ARi5A)

**Comment by erictraut** [comment]
OK, thanks. I'm able to repro that, and I'll investigate further. Apologies for the regression.

**Comment by erictraut** [comment]
The issue was a long-standing bug that was exposed by a recent change I made. It's specific to dataclasses and `type` or `type[Any]`. We didn't have any test cases that specifically tested that combination.

The `hydra-zen` library is part of our regression tests (via the `mypy-primer` tool), so I was curious why I didn't catch this regression through that mechanism. Here's the mypy-primer results from the commit that exposed the regression: https://github.com/microsoft/pyright/pull/6553#issuecomment-1826474399. You can see that there were a few small differences reported for `hydra-zen`, but they were all new "Unnecessary "# type: ignore" comment" errors. I interpreted this to mean that you had worked around a pyright bug using a `# type: ignore` comment, and the workaround was no longer needed because of the bug fix I was introducing. Let me know if I got that analysis wrong.

In any event, this regression will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.339, which I just published. It will also be included in a future release of pylance.
